error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/StreamRules.java

```diff
@@ -250,6 +256,99 @@ Optional<S> after(Stream<T> stream, Function<? super T, S> function) {
+    }
+
+    @AfterTemplate
+    boolean after(Stream<T> stream) {
+      return stream.findAny().isEmpty();
+    }
+  }
+
+  // XXX: This rule assumes that any matched `Collector` does not perform any filtering.
+  static final class StreamCollectingAndThenImmutableListIsEmpty<T> {
+    @BeforeTemplate
+    boolean before(
+        Stream<T> stream, Collector<? super T, ?, ? extends ImmutableList<?>> collector) {
+      return stream.collect(collectingAndThen(collector, ImmutableList::isEmpty));
+    }
+
+    @AfterTemplate
+    boolean after(Stream<T> stream) {
+      return stream.findAny().isEmpty();
+    }
+  }
+
+  // XXX: This rule assumes that any matched `Collector` does not perform any filtering.
+  static final class StreamCollectingAndThenListIsEmpty<T> {
+    @BeforeTemplate
+    boolean before(Stream<T> stream, Collector<? super T, ?, ? extends List<?>> collector) {
+      return stream.collect(collectingAndThen(collector, List::isEmpty));
+    }
+
+    @AfterTemplate
+    boolean after(Stream<T> stream) {
+      return stream.findAny().isEmpty();
+    }
+  }
+
+  // XXX: This rule assumes that any matched `Collector` does not perform any filtering.
+  static final class StreamCollectingAndThenImmutableSetIsEmpty<T> {
+    @BeforeTemplate
+    boolean before(Stream<T> stream, Collector<? super T, ?, ? extends ImmutableSet<?>> collector) {
+      return stream.collect(collectingAndThen(collector, ImmutableSet::isEmpty));
+    }
+
+    @AfterTemplate
+    boolean after(Stream<T> stream) {
+      return stream.findAny().isEmpty();
+    }
+  }
+
+  // XXX: This rule assumes that any matched `Collector` does not perform any filtering.
+  static final class StreamCollectingAndThenSetIsEmpty<T> {
+    @BeforeTemplate
+    boolean before(Stream<T> stream, Collector<? super T, ?, ? extends Set<?>> collector) {
+      return stream.collect(collectingAndThen(collector, Set::isEmpty));
+    }
+
+    @AfterTemplate
+    boolean after(Stream<T> stream) {
+      return stream.findAny().isEmpty();
+    }
+  }
+
+  // XXX: This rule assumes that any matched `Collector` does not perform any filtering.
+  static final class StreamCollectingAndThenImmutableMapIsEmpty<T> {
+    @BeforeTemplate
+    boolean before(
+        Stream<T> stream, Collector<? super T, ?, ? extends ImmutableMap<?, ?>> collector) {
+      return stream.collect(collectingAndThen(collector, ImmutableMap::isEmpty));
+    }
+
+    @AfterTemplate
+    boolean after(Stream<T> stream) {
+      return stream.findAny().isEmpty();
+    }
+  }
+
+  // XXX: This rule assumes that any matched `Collector` does not perform any filtering.
+  static final class StreamCollectingAndThenMapIsEmpty<T> {
+    @BeforeTemplate
+    boolean before(Stream<T> stream, Collector<? super T, ?, ? extends Map<?, ?>> collector) {
+      return stream.collect(collectingAndThen(collector, Map::isEmpty));
+    }
+
+    @AfterTemplate
+    boolean after(Stream<T> stream) {
+      return stream.findAny().isEmpty();
```

Comment:
W.r.t. your third question: rules are named after their `@AfterTemplate`. For new templates we'd name this rule `StreamFindAnyIsEmpty`. However, there is already a `StreamIsEmpty` template below with exactly this same `@AfterTemplate` definition, so let's try to merge the changes into that one.